### PR TITLE
Single identifier type

### LINKED ISSUE #482: Identifier, Regex/Multi/Exact enum

### PR DESCRIPTION
**Pull request by psnairne**
As it was months ago, there is once again the single variant of Identifier

### CODE REVIEW

File: integration_tests/tests/assets/integration_test/input_data/csv_data_4.csv [hunk 1]
```diff
@@ -1,4 +1,4 @@
-Patient ID, diseases, disease_onset, gene, hgvs1, hgvs2
+Patient ID,diseases,disease_onset,gene,hgvs1,hgvs2
```
Comment (psnairne) [inline integration_tests/tests/assets/integration_test/input_data/csv_data_4.csv:1]: This was needed because headers are now matched exactly, spaces aren't allowed

File: phenoxtract/src/config/table_context.rs [hunk 2]
```diff
@@ -138,14 +144,20 @@ impl From<&Regex> for Identifier {
     }
 }
 
+impl Identifier {
+    pub fn regex_from_str(value: &str) -> Identifier {
```
Comment (SmartMonkey-git) [inline phenoxtract/src/config/table_context.rs]: This should probably return a result. Not sure, if it should be a ConstructionError. For ease we can use impl Into<String> here.

```suggestion
    pub fn regex_from_str(value: impl Into<String>) -> Result<Identifier, ConstructionError> {
```

File: phenoxtract/src/config/table_context.rs [hunk 3]
```diff
@@ -169,13 +181,14 @@ impl From<&[String]> for Identifier {
...
+        Identifier::from("")
     }
 }
 
 impl Display for Identifier {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
         match self {
+            Identifier::Single(single) => write!(f, "{}", single),
             Identifier::Regex(regex) => write!(f, "{}", regex),
             Identifier::Multi(multi) => write!(f, "{}", multi.join(".")),
```
Comment (SmartMonkey-git) [inline phenoxtract/src/config/table_context.rs]:
```suggestion
            Identifier::Single(single) => write!(f, "Single({})", single),
Identifier::Regex(regex) => write!(f, "Regex({})", regex),
Identifier::Multi(multi) => write!(f, "Multi({})", multi.join(".")),
```

File: phenoxtract/src/config/table_context.rs [hunk 4]
```diff
@@ -185,23 +198,16 @@ impl Display for Identifier {
...
+                .filter(|col| *col == id.as_str())
+                .collect(),
+            Identifier::Regex(id) => col_names
+                .iter()
+                .copied()
+                .filter(|col| id.is_match(col))
+                .collect(),
             Identifier::Multi(ids) => col_names
                 .iter()
                 .copied()
```
Comment (SmartMonkey-git) [inline phenoxtract/src/config/table_context.rs]: Lets only clone col_names that match.

```
        match self {
            Identifier::Single(id) => col_names
                .iter()
                .filter(|col| col == &&id.as_str())
                .cloned()
                .collect(),
            Identifier::Regex(id) => col_names
                .iter()
                .filter(|col| id.is_match(col))
                .cloned()
                .collect(),
            Identifier::Multi(ids) => col_names
                .iter()
                .filter(|col| ids.iter().any(|id| id == *col))
                .cloned()
                .collect(),
        }
```

File: phenoxtract/src/config/table_context.rs [hunk 5]
```diff
@@ -406,22 +412,22 @@ mod tests {
     }
 
     #[test]
-    fn regex_exact_match_returns_match() {
+    fn single_id_returns_match() {
```
Comment (SmartMonkey-git) [inline phenoxtract/src/config/table_context.rs:420]: We should add the edge case, we found in the i data here as a separate test.

**Review by SmartMonkey-git — COMMENTED** [review]
Some comments

File: phenoxtract/src/config/table_context.rs (hunk 4 above)
Comment (psnairne) [reply phenoxtract/src/config/table_context.rs]: good spot

File: phenoxtract/src/validation/table_context_validation.rs [hunk 6]
```diff
@@ -14,12 +14,15 @@ pub(crate) fn validate_unique_identifiers(
     series_contexts
         .iter()
         .for_each(|sc| match sc.get_identifier() {
+            Identifier::Single(s) => {
```
Comment (SmartMonkey-git) [inline phenoxtract/src/validation/table_context_validation.rs:18]: Here is a cleaner version that needs no cloning:

```
pub(crate) fn validate_unique_identifiers(
    series_contexts: &[SeriesContext],
) -> Result<(), ValidationError> {
    let mut unique_identifiers: HashSet<&str> = HashSet::new();
    let mut duplicates: Vec<&str> = Vec::new();

    for sc in series_contexts.iter() {
        let ids: Vec<&str> = match sc.get_identifier() {
            Identifier::Single(s) => vec![s.as_str()],
            Identifier::Regex(regex) => vec![regex.as_str()],
            Identifier::Multi(multi_ids) => multi_ids.iter().map(|id| id.as_str()).collect(),
        };
        for id in ids {
            if !unique_identifiers.insert(id) {
                duplicates.push(id);
            }
        }
    }

    fail_validation_on_duplicates(
        &duplicates,
        "duplicates",
        "Found duplicate identifiers in SeriesContexts",
    )
}
```

**Review by SmartMonkey-git — APPROVED** [review]

File: phenoxtract/src/validation/table_context_validation.rs (hunk 6 above)
Comment (psnairne) [reply phenoxtract/src/validation/table_context_validation.rs:18]: I'm not actually sure how much sense this validation function makes anymore. Since if a Regex identifier is z{3} that means something different to a column called z{3}, in fact, they wouldn't match.
